src/main/java/reposense/authorship/analyzer/AnnotatorAnalyzer.java

```diff
@@ -23,43 +24,46 @@
      * Overrides the authorship information in {@code fileInfo} based on annotations given on the file.
      */
     public static void aggregateAnnotationAuthorInfo(FileInfo fileInfo, Map<String, Author> authorAliasMap) {
-        Author currentAuthor = Author.UNKNOWN_AUTHOR;
+        Optional<Author> currentAnnotatedAuthor = Optional.empty();
         Path filePath = Paths.get(fileInfo.getPath());
         for (LineInfo lineInfo : fileInfo.getLines()) {
             if (lineInfo.getContent().contains(AUTHOR_TAG)) {
-                Author newAuthor = findAuthorInLine(lineInfo.getContent(), authorAliasMap);
+                Optional<Author> newAnnotatedAuthor = findAuthorInLine(lineInfo.getContent(), authorAliasMap);
 
-                if (newAuthor.equals(Author.UNKNOWN_AUTHOR)) {
-                    //end of an author tag should belong to this author too.
-                    lineInfo.setAuthor(currentAuthor);
-                } else if (newAuthor.getIgnoreGlobMatcher().matches(filePath)) {
-                    newAuthor = Author.UNKNOWN_AUTHOR;
+                if (!newAnnotatedAuthor.isPresent()) {
+                    // if an end tag is used (i.e. @@author), attribute this line to the
+                    // author found in the start tag (i.e. the currentAnnotatedAuthor)
+                    // if an end tag was used without a corresponding start tag, attribute the
+                    // line to UNKNOWN_AUTHOR
+                    lineInfo.setAuthor(currentAnnotatedAuthor.orElse(Author.UNKNOWN_AUTHOR));
+                } else if (newAnnotatedAuthor.get().getIgnoreGlobMatcher().matches(filePath)) {
+                    newAnnotatedAuthor = Optional.empty();
                 }
 
                 //set a new author
-                currentAuthor = newAuthor;
-            }
-            if (!currentAuthor.equals(Author.UNKNOWN_AUTHOR)) {
-                lineInfo.setAuthor(currentAuthor);
+                currentAnnotatedAuthor = newAnnotatedAuthor;
             }
+            currentAnnotatedAuthor.ifPresent(lineInfo::setAuthor);
         }
     }
 
     /**
      * Extracts the author name from the given {@code line}, finds the corresponding {@code Author}
-     * in {@code authorAliasMap},and returns the result.
-     * @return {@code Author#UNKNOWN_AUTHOR} if no matching {@code Author} is found.
+     * in {@code authorAliasMap}, and returns this {@code Author} stored in an {@code Optional}.
+     * @return {@code Optional.of(Author#UNKNOWN_AUTHOR)} if no matching {@code Author} is found,
+     *         {@code Optional.empty()} if an end author tag is used (i.e. "@@author"),
+     *         and {@code Optional.empty()} if the extracted author name is too short.
```

Comment:
Why return `Optional.empty()` if the extracted author name is too short?

Reply:
**In the original codebase**, the extracted author name is validated against this regex: 
`[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]`

If the extracted author name does not match this regex, then the author is set to `UNKNOWN_AUTHOR` and the annotation override is not applied to the code block.  This regex ensures the author name has at least 2 consecutive alphanumeric characters.

In order to keep this behaviour, I chose to return `Optional.empty()` when the extracted author name fails the regex check.

I am returning `Optional.of(UNKNOWN_AUTHOR)` in cases where we detect the author wants to disown the code block (by using a `@@author random-user` tag)

Reply:
Hmmm... Regarding the regex, I suppose the main intention of the founding developer who adopted this regex is to conform to the github's GitHub ID.

FYI, GitHub ID requirement is:
* Max 39 characters long
* Username must me alphanumeric and can contain up to a single hyphen, but username cannot begin or end with a hyphen

However the thing is, authors can use any type of character that they want when they commit in git, such as this
![image](https://user-images.githubusercontent.com/35756209/75038387-3f0eaf80-54f1-11ea-8c71-2f6a841aacba.png)

Since we are relying on Git author name to attribute code authorship, I think it's better if we change/remove the author regex to suit this current situation. You might want to create an issue on this and/or put this on hold for now.

If you want a better understanding on this, try this:

* use an author tag with an author name that doesn't conform to github id convention (e.g.: `-my author name-`)
* Run the program

Reply:
Thanks for the thoughts @jylee-git 
Yes, we could update our regex to be more flexible (allow things like `@@author --test--`)
I'll open up an issue on this.

What about this PR itself? Do you mean this PR should be placed on hold?
What are your thoughts on returning `Optional.empty()` for cases where the regex validation fails?

Reply:
I think you can continue fixing this bug.

One more thing to ask: is the current intended solution, when a person does not want to attribute the code to anyone, to use an author name that hasn't been used to commit the code before?
> e.g.: the git id `some-randomUser` did not make any commits to the code before. In order to **not** attribute the code to anyone, the person will use `//@@author some-randomUser` to avoid attributing the code below to anyone.

Reply:
Yes, that is the intended solution (as mentioned by Prof @damithc [here](https://github.com/reposense/RepoSense/issues/823#issuecomment-513143467))